clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/TaskHealthCachingAgent.java

```diff
@@ -299,34 +289,52 @@ private TaskHealth inferHealthNetworkBindedContainer(
     return overallTaskHealth;
   }
 
+  private TargetHealthDescription findHealthDescription(
+      List<TargetHealthDescription> targetHealths, String targetId, Integer targetPort) {
+
+    return targetHealths.stream()
+        .filter(
+            h ->
+                h.getTarget().getId().equals(targetId)
+                    && h.getTarget().getPort().equals(targetPort))
+        .findFirst()
+        .orElse(null);
+  }
+
   private TaskHealth describeTargetHealth(
-      AmazonElasticLoadBalancing amazonloadBalancing,
+      TargetHealthCacheClient targetHealthCacheClient,
       Task task,
       Service loadBalancerService,
       String serviceName,
       String targetGroupArn,
       String targetId,
       Integer targetPort,
       TaskHealth overallTaskHealth) {
-    DescribeTargetHealthResult describeTargetHealthResult =
-        amazonloadBalancing.describeTargetHealth(
-            new DescribeTargetHealthRequest()
-                .withTargetGroupArn(targetGroupArn)
-                .withTargets(new TargetDescription().withId(targetId).withPort(targetPort)));
-
-    if (describeTargetHealthResult.getTargetHealthDescriptions().isEmpty()) {
-      log.debug("Target health description is empty");
+
+    String targetHealthKey = Keys.getTargetHealthKey(accountName, region, targetGroupArn);
+    EcsTargetHealth targetHealth = targetHealthCacheClient.get(targetHealthKey);
+
+    if (targetHealth == null) {
```

Comment:
Is there a possibility that the cache doesn't contain an entry but the DescribeHealthResult would show that the task is unhealthy? (i.e. the cache is a bit stale, hasn't pulled/updated recently). Would it make sense to call describeTargetHealth if the cache doesn't have an entry, and then return if that is also null?

Also, not sure if we really need to evict here if the cache is always getting everything (on pull).

Reply:
In the scenario mentioned above, if a new task was started as part of a deployment but was not added to the cache, this method would return success if any one of the tasks yet to start up was not in the cache.

Reply:
it's possible it may not find the given target (due to staleness), and right now the approach is to wait for the downstream caches to catch up and declare the task healthy at that point. This applies to the the other ECS caches called here as well. 

While I like the immediacy of calling ELB as a fallback, I'm worried this might be invoked too frequently to retain most of the benefit of the dedicated TargetHealth cache. My preference would be to see how this goes and incorporate additional/fallback logic if increased wait times for "up instances" become a problem (anecdotally, I deployed a few times and didn't see a noticeable increase). However, my hunch is that rate limiting is the bigger bottleneck here ATM.

Reply:
> In the scenario mentioned above, if a new task was started as part of a deployment but was not added to the cache, this method would return success if any one of the tasks yet to start up was not in the cache.

You're right that if the task has multiple target groups attached then this would return "healthy" if one was found but other's were empty (b/c it's returning the previous overallHealth, not "unknown"). I'll make that change and test it out, then update this.

Reply:
Awesome! Thanks @allisaurus!